### PR TITLE
add MAD outlier rule to find outliers

### PR DESCRIPTION
**Pull request by GLevV**
PR for #496 
Added MAD outlier rule
Refactored base code and tests

Extend outlier detection functionality of OutlierTrimmer and Winsorizer.

### CODE REVIEW

**Review by solegalli — COMMENTED** [review]
Hi @GLevV 

Really, really good implementation of the outlier handlers. Thank you very much for taking the time to re-write almost the entire logic. Very elegant. I appreciate it a lot!

I also learned a lot.

There are a couple of tests failing. 

Stylechecks is complaining about numpy imported but being unused, see below:

```
stylechecks run-test: commands[0] | flake8 feature_engine tests
feature_engine/outliers/base_outlier.py:3:1: F401 'numpy as np' imported but unused
feature_engine/outliers/base_outlier.py:179:89: E501 line too long (89 > 88 characters)
feature_engine/outliers/base_outlier.py:195:89: E501 line too long (91 > 88 characters)
feature_engine/outliers/base_outlier.py:218:89: E501 line too long (90 > 88 characters)
feature_engine/outliers/trimmer.py:4:1: F401 'numpy as np' imported but unused
tests/test_outliers/test_outlier_trimmer.py:4:1: F401 'numpy as np' imported but unused
```

The Python tests are failing because of the error to catch when scale 

This test is also failing:
https://github.com/feature-engine/feature_engine/blob/65fedf100bc85f364b26838866a264dbf7320841/tests/test_outliers/test_winsorizer.py#L303

I believe they should all be small fixes. Would you be able to have a look at it?

I think once the tests pass this is ready to go!

Thank you very much!

File: feature_engine/outliers/base_outlier.py [hunk 1]
```diff
@@ -259,42 +266,46 @@ def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None):
...
+            bias = X[self.variables_].quantile((0.75, 0.25))
+            scale = bias.loc[0.75] - bias.loc[0.25]
+        elif self.capping_method == "quantiles":
+            bias = X[self.variables_].quantile((1 - self.fold, self.fold))
+            scale = bias.loc[1 - self.fold] - bias.loc[self.fold]
+        elif self.capping_method == "mad":
+            bias = X[self.variables_].median()
+            # scaling factor for normal distribution
+            scale = (X[self.variables_] - bias).abs().median() / 0.67449
+        if (scale == 0).any():
```
[the remark this answers is not in the document]
Comment (GLevV) [reply feature_engine/outliers/base_outlier.py:282]: Should it be added to BaseOutlier class or to Winsorizer and Trimmer separately?

**Comment by GLevV** [comment]
Fixes implemented

**Comment by solegalli** [comment]
Hey @GLevV 

This implementation is great. Thank you so much!

Code-wise, this is good to go.

If I may ask a little bit more from you, could you please add the Mad rule to the user guide? In this 2 files:

https://github.com/feature-engine/feature_engine/blob/main/docs/user_guide/outliers/Winsorizer.rst

https://github.com/feature-engine/feature_engine/blob/main/docs/user_guide/outliers/OutlierTrimmer.rst

Thank you!

**Pull request merged**